lib/galaxy/tool_util/xsd/galaxy.xsd

```diff
@@ -1314,7 +1314,12 @@ When this attribute is true and ``compare`` is set to ``diff``, try to decompres
     </xs:attribute>
     <xs:attribute name="delta" type="xs:integer">
       <xs:annotation>
-        <xs:documentation xml:lang="en">If ``compare`` is set to ``sim_size``, this is the number of bytes different allowed (default 10000).</xs:documentation>
+        <xs:documentation xml:lang="en">If ``compare`` is set to ``sim_size``, this is the number of bytes different allowed. Can be combined with ``delta_frac``</xs:documentation>
+      </xs:annotation>
+    </xs:attribute>
+    <xs:attribute name="delta_frac" type="xs:float">
+      <xs:annotation>
+        <xs:documentation xml:lang="en">If ``compare`` is set to ``sim_size``, the size of the smaller file in bytes has to be at least this fraction times the size of the larger file in bytes. Can be combined with ``delta``. (default 1.0).</xs:documentation>
```

Comment:
I find the text hard to understand. I guess the point is to allow a margin ? Since this seems independent of the delta I wouldn't mention it and rename this to `size_margin`. Which file is bigger doesn't matter for a margin, so I think that would be easier to understand.


```suggestion
    <xs:attribute name="size_margin" type="xs:float" default="0">
      <xs:annotation>
        <xs:documentation xml:lang="en">If ``compare`` is set to ``sim_size`` the test will fail if the difference in output size exceeds the test data output by ``size_margin``. 0 means no margin, 0.1 means 10% margin, 2 means the output may be 2 times larger or smaller than the provided test output dataset. Note that ``sim_size`` is applied independently. </xs:documentation>
```

Reply:
I like the idea of giving examples and I agree that the original formulation has problems.

But there are some words in your formulation that are wrong or might be misleading:

- actually the value of `delta_frac` has to be between 0 and 1 (larger values work, but are meaningless since 1 already allows any difference)
- In your example a value of 0.1 would allow the size differ by a factor of 10, and a 10% size difference should correspond to a delta_frac=0.9
- "difference" indicates (to me) that the error margin is additive 
- "exceeds the test data output" actually its the lager of the test file and the test output
- Also note that the default is actually 1.

> Note that ``sim_size`` is applied independently

actually one needs to set ``sim_size``, its `delta` that is applied independently.

We can still think about the semantics that are easiest to understand. Currently the test is `min/max > delta_frac`, i.e. `min > delta_frac * max`. 

But we could as well invert this `max/min > delta_frac`. Maybe even `log(min/max)>delta_frac` (which makes expressing orders of magnitude easy).

Few comments from you would be great. Then I will work on the formulation.

Reply:
I think I've changed my mind, `margin` isn't well-defined, `delta_frac` is better.
How about
```suggestion
    <xs:attribute name="delta_frac" type="xs:float">
      <xs:annotation>
        <xs:documentation xml:lang="en">If ``compare`` is set to ``sim_size``, the relative difference in size allowed between the generated output and the provided test dataset. 0 means no difference in size allowed, 0.1 means 10% of difference allows, 2 means that the difference may be twice the size of the provided test dataset. Note that ``delta`` is applied independently. (default 0).</xs:documentation>
```

The implementation then could be (assuming s1 is the provided test dataset, not sure on that)
```python
actual_delta = abs(s1 - s2)
actual_delta_frac = actual_delta / s1
if actual_delta_frac > delta_frac:
    raise AssertionError
```

Reply:
I would prefer the current implementation over your suggestion, because it appears to be (more) analogous to the implementation for `delta` (`abs(s1 - s2)`), i.e. it just tests for the absolute size difference and it does not matter which file is the provided test dataset and which the test output.

Reply:
If you can rewrite the help to be more understandable and calculable by a user I can get on board with that. Which file is the provided file should be easy to figure out.

Reply:
Inverted the meaning of the value: `min/max > delta` got `1-min/max>delta` (which is equivalent to `abs(s1/s2)/max>delta`).

This is nearly what you meant previously (`actual_delta_frac = actual_delta / s1`) .. I now see the advantage that both values allow to make a statement on the same thing (the abs difference) .. just in different ways.

Reply:
It does read better now, but  I think it's inconvenient that the meaningful values max out at 1.